### PR TITLE
Provide two-stage interface to front module

### LINKED ISSUE #451: Provide two-stage interface to front module
In order to support #438, we need to provide a two stage interface to the front module:

1. In the first stage, it provides a list of the module’s imports. In this stage, it runs the parser/tokenizer only up till the end of the import declarations. It doesn’t yet build a complete AST.
2. In the second stage, it generates a BIR module; it is provided with the public interface to each of the modules that it imports. In this stage , it builds the complete AST.

### PR DESCRIPTION
**Pull request by manuranga**
Provide two-stage interface to front module
Fixes  #451

### CODE REVIEW

File: compiler/modules/front.syntax/parse.bal [hunk 1]
```diff
@@ -7,20 +7,44 @@ public type FilePath record {|
...
+        return imports;
+    }
+
+    public function parse() returns ModulePart|err:Syntax {
+        Tokenizer? t = self.tok;
+        if t is () {
+            panic error("parseImports() must be called before parse()");
+        }
+
+        Tokenizer tok = <Tokenizer>t;
```
Comment (manuranga) [inline compiler/modules/front.syntax/parse.bal]: Is there a better way do this?

File: compiler/modules/front/front.bal [hunk 2]
```diff
@@ -133,20 +133,46 @@ type LoadedSourcePart record {|
...
+
+    function init(s:Parser[] parsers,s:ImportDecl[] imports , bir:ModuleId id) {
+        self.parsers = parsers;
+        self.imports = imports;
+        self.id = id;
+    }
+
+    function getImports() returns s:ImportDecl[] => self.imports;
+
+    public function load(t:Env env) returns ResolvedModule|err:Any|io:Error {
```
Comment (jclark) [inline compiler/modules/front/front.bal]: Call this `resolve`.

File: compiler/modules/front.syntax/parse.bal (hunk 1 above)
Comment (jclark) [reply compiler/modules/front.syntax/parse.bal]: Move parseImports back out of Parser. Call parseImports before you construct the parser. Construct the parser with a non-nil Tokenizer and a ModulePart that contains the list of imports. Provide a function getImports() that gets the imports from the module part.

Parser is not quite the right word. It's more like PartiallyParsedModule, but that's too long. Perhaps you can think of a better word.

Comment (jclark) [reply compiler/modules/front.syntax/parse.bal]: I think it's maybe not too difficult to make this readonly.

You can save Tokenizer state just by doing cloneReadOnly().  Then you just need to add a method on Tokenizer `function restore(Tokenizer t)` which resets the non-final fields from `t` (panicking if the final fields are not `===`). I expect we will need something like this for #412.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front.syntax/parse.bal (hunk 1 above)
Comment (jclark) [reply compiler/modules/front.syntax/parse.bal:42]: Actually there's no need for getImports. Just don't make `part` private. We are not trying to encapsulate parsing in this class.

Comment (jclark) [reply compiler/modules/front.syntax/parse.bal]: Should be `getImportDecls()`

Comment (manuranga) [reply compiler/modules/front.syntax/parse.bal]: >You can save Tokenizer state just by doing cloneReadOnly(). 

I am not sure I understand, I though in Ballerina `cloneReadOnly()` doesn't work on objects.

Should I add methods to Tokenizer to dump state `function state() TokenizerState` and restore form it `function restore(TokenizerState s)`?

Comment (manuranga) [reply compiler/modules/front.syntax/parse.bal]: To clarify:

```ballerina
    Tokenizer tok = new (file);
    readonly & Tokenizer t = tok.cloneReadOnly()
```
gives
```
undefined method 'cloneReadOnly' in object 'wso2/nballerina.front.syntax:0.1.0:Tokenizer'
```

Comment (manuranga) [reply compiler/modules/front.syntax/parse.bal]: For the record, we decided to add a method to save state.

**Comment by manuranga** [comment]
I've run into an issue where I can't assign `SourceFile` to a `readonly & SourceFile` type variable. I expect this to be possible given that `SourceFile` class itself is readonly. Is this a jballerina bug?

**Comment by jclark** [comment]
> I've run into an issue where I can't assign `SourceFile` to a `readonly & SourceFile` type variable. I expect this to be possible given that `SourceFile` class itself is readonly. Is this a jballerina bug?

Sounds like it: `SourceFile` and `readonly & SourceFile` are the same type. I've run into this sort of thing before. Sometimes you can workaround by introducing a separate temp variable.

Why do you need `readonly & SourceFile` type variable? If you let me see the code, I can suggest some possible workarounds.

**Comment by jclark** [comment]
If you cannot get the readonly side working, send the pull request without this and I will have a go.

File: compiler/modules/front.syntax/parse.bal [hunk 3]
```diff
@@ -7,20 +7,49 @@ public type FilePath record {|
...
+// xxx alternative names ModulePartPrecursor ModulePrecursor ResolvableModulePart ScannedModulePart
+public class PreModulePart {
+    private TokenizerState tokState;
+    private ImportDecl[] imports;
+    private SourceFile file;
+    private int partIndex;
+
+    function init(SourceFile file, readonly & ImportDecl[] imports, TokenizerState tokState, int partIndex) {
+        // xxx JBUG when class is readonly, gives error `incompatible types`
+        self.file = file;
```
Comment (manuranga) [inline compiler/modules/front.syntax/parse.bal:23]: @jclark This is where I face the issue.

File: compiler/modules/front.syntax/token.bal [hunk 4]
```diff
@@ -92,6 +92,16 @@ function toToken(string t) returns Token {
...
 }
 
+type TokenizerState readonly & record {
+    int lineIndex;
+    int fragCodeIndex;
+    int codePointIndex;
+    int fragmentIndex;
+    int tokenStartCodePointIndex;
+    Mode mode;
+    Token? curTok;
```
Comment (manuranga) [inline compiler/modules/front.syntax/token.bal:103]: Another options is to have `Token curTok;` and change `TokenizerState tokState` to `TokenizerState? tokState` in `PreModulePart`'s `init`.

File: compiler/modules/front.syntax/parse.bal [hunk 5]
```diff
@@ -7,20 +7,49 @@ public type FilePath record {|
...
-        importDecl: check parseImportDecl(tok)
-    };
-    while tok.current() != () {
-        part.defns.push(check parseModuleDecl(tok, part));
+    // xxx inline after #443, and check if cloneReadOnly can be avoided
+    ImportDecl? im = check parseImportDecl(tok);
+    return new(file, im == () ? [] : [im.cloneReadOnly()], tok.save(), partIndex);
+}
+
+// xxx alternative names ModulePartPrecursor ModulePrecursor ResolvableModulePart ScannedModulePart
```
Comment (jclark) [inline compiler/modules/front.syntax/parse.bal]: I think ScannedModulePart is the best. Then you can call the function scanModulePart.

File: compiler/modules/front.syntax/token.bal [hunk 6]
```diff
@@ -275,6 +285,30 @@ class Tokenizer {
         return err:syntax(msg, loc=err:location(self.file, self.currentPos()));
     }
 
+    function save() returns TokenizerState {
+        return { lineIndex : self.lineIndex,
```
Comment (jclark) [inline compiler/modules/front.syntax/token.bal]: start lineIndex on a separate line

File: compiler/modules/front.syntax/token.bal [hunk 7]
```diff
@@ -275,6 +285,30 @@ class Tokenizer {
...
+                 fragCodeIndex : self.fragCodeIndex,
+                 codePointIndex : self.codePointIndex,
+                 fragmentIndex : self.fragmentIndex,
+                 tokenStartCodePointIndex : self.tokenStartCodePointIndex,
+                 mode : self.mode,
+                 curTok : self.curTok.cloneReadOnly() };
```
Comment (jclark) [inline compiler/modules/front.syntax/token.bal]: Better to make Token be readonly, which requires changing VariableLengthToken.

File: compiler/modules/front.syntax/token.bal (hunk 4 above)
Comment (jclark) [reply compiler/modules/front.syntax/token.bal:103]: What you have seems fine to me.

File: compiler/modules/front/front.bal [hunk 8]
```diff
@@ -133,20 +133,46 @@ type LoadedSourcePart record {|
...
+        s:ScannedModulePart p = check s:scanModulePart(loaded.lines, loaded.path, i);
+        parts.push(p);
+        imports.push(... check p.getImportDecls());
+    }
+    ScannedModule mod = new(parts, imports, id);
+    return mod;
+}
+
+public class ScannedModule {
+    private s:ScannedModulePart[] parsers;
```
Comment (manuranga) [inline compiler/modules/front/front.bal]: Need to rename this.

Comment (manuranga) [reply compiler/modules/front/front.bal]: fixed

File: compiler/modules/front.syntax/parse.bal (hunk 3 above)
Comment (jclark) [reply compiler/modules/front.syntax/parse.bal:23]: Try declaring `lines` in `SourceFile` as `readonly & ScannedLine[]`.

Comment (manuranga) [reply compiler/modules/front.syntax/parse.bal:23]: This didn't work, but found a different workaround below, please take a look.

File: compiler/modules/front.syntax/token.bal (hunk 7 above)
Comment (manuranga) [reply compiler/modules/front.syntax/token.bal]: fixed

File: compiler/modules/front.syntax/token.bal [hunk 9]
```diff
@@ -275,14 +285,50 @@ class Tokenizer {
...
+    }
+
 }
 
 function createPosition(int line, int column) returns Position {
     return (line << 32) | column;
 }
 
-public readonly class SourceFile {
+// JBUG when SourceFile is replaced with the ScannedSourceFile, get `incompatible types`
```
Comment (manuranga) [inline compiler/modules/front.syntax/token.bal:321]: This is the workaround, but I got some errors when I tried to use type inclusions. But it still works without that.

File: compiler/modules/front.syntax/parse.bal [hunk 10]
```diff
@@ -7,20 +7,49 @@ public type FilePath record {|
...
+    public function getImportDecls() returns ImportDecl[]|err:Syntax {
+        return self.imports;
+    }
+
+    public function parse() returns ModulePart|err:Syntax {
+        Tokenizer tok = new(self.file);
+        tok.restore(self.tokState);
+        ModuleLevelDefn[] defns = [];
+        // inline after #443
+        ImportDecl? im = self.imports.length() > 0 ? self.imports[0] : ();
```
Comment (jclark) [inline compiler/modules/front.syntax/parse.bal]: Prefer `decl` to `im`

File: compiler/modules/front.syntax/token.bal [hunk 11]
```diff
@@ -275,14 +285,50 @@ class Tokenizer {
...
+public type SourceFile readonly & object {
+    // xxx should err:File be included here ?
```
Comment (jclark) [inline compiler/modules/front.syntax/token.bal]: Yes, you should include err:File

File: compiler/modules/front/front.bal [hunk 12]
```diff
@@ -133,20 +133,46 @@ type LoadedSourcePart record {|
...
+    ScannedModule mod = new(parts, imports, id);
+    return mod;
+}
+
+public class ScannedModule {
+    private s:ScannedModulePart[] scannedParts;
+    private s:ImportDecl[] imports;
+    private bir:ModuleId id;
+
+    function init(s:ScannedModulePart[] scannedParts,s:ImportDecl[] imports , bir:ModuleId id) {
```
Comment (jclark) [inline compiler/modules/front/front.bal]: Couple of spaces wrong

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front.syntax/token.bal (hunk 11 above)
Comment (manuranga) [reply compiler/modules/front.syntax/token.bal]: it's giving me `object type inclusion cannot be used with a 'readonly' type descriptor in an 'object' type descriptor` error. JBUG?

Comment (jclark) [reply compiler/modules/front.syntax/token.bal]: Probably. Let's get this in and I will try to find a better workaround.

File: compiler/modules/front.syntax/parse.bal [hunk 13]
```diff
@@ -34,7 +59,8 @@ public function parseExpression(string[] lines, FilePath path) returns Expr|err:
 }
 
 function createSourceFile(FilePath path, string[] lines) returns SourceFile {
-    return new(path, scanLines(lines));
+    ScannedSourceFile x = new(path, scanLines(lines));
```
Comment (manuranga) [inline compiler/modules/front.syntax/parse.bal]: not a good name, will change

**Pull request merged**